pkg/controller/autoops/api_key.go

```diff
@@ -0,0 +1,403 @@
+	_, err := esClient.InvalidateAPIKeys(ctx, esclient.APIKeysInvalidateRequest{
+		IDs: []string{keyID},
+	})
+	return err
+}
+
+// storeAPIKeyInSecret stores the API key in a Kubernetes secret.
+func storeAPIKeyInSecret(
+	ctx context.Context,
+	c k8s.Client,
+	encodedKey string,
+	expectedHash string,
+	policy autoopsv1alpha1.AutoOpsAgentPolicy,
+	es esv1.Elasticsearch,
+) error {
+	secretName := apiKeySecretNameFor(types.NamespacedName{Namespace: policy.Namespace, Name: es.Name})
+	expected := buildAutoOpsESAPIKeySecret(policy, es, secretName, encodedKey, expectedHash)
+
+	reconciled := &corev1.Secret{}
+	return reconciler.ReconcileResource(
+		reconciler.Params{
+			Context:    ctx,
+			Client:     c,
+			Owner:      &policy,
+			Expected:   &expected,
+			Reconciled: reconciled,
+			NeedsUpdate: func() bool {
+				return !maps.IsSubset(expected.Labels, reconciled.Labels) ||
+					!maps.IsSubset(expected.Annotations, reconciled.Annotations) ||
+					!reflect.DeepEqual(expected.Data, reconciled.Data)
+			},
+			UpdateReconciled: func() {
+				reconciled.Labels = maps.Merge(reconciled.Labels, expected.Labels)
+				reconciled.Annotations = maps.Merge(reconciled.Annotations, expected.Annotations)
+				reconciled.Data = expected.Data
+			},
+		},
+	)
+}
+
+// buildAutoOpsESAPIKeySecret builds the expected Secret for autoops ES API key.
+func buildAutoOpsESAPIKeySecret(policy autoopsv1alpha1.AutoOpsAgentPolicy, es esv1.Elasticsearch, secretName string, encodedKey string, expectedHash string) corev1.Secret {
+	meta := metadata.Propagate(&policy, metadata.Metadata{
+		Labels:      newMetadataFor(&policy, &es, expectedHash),
+		Annotations: policy.GetAnnotations(),
+	})
+
+	// The 'managed-by' is only needed/wanted for the API key in
+	// Elasticsearch so we remove it from the secret.
+	delete(meta.Labels, managedByMetadataKey)
+
+	return corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        secretName,
+			Namespace:   policy.GetNamespace(),
+			Labels:      meta.Labels,
+			Annotations: meta.Annotations,
+		},
+		Data: map[string][]byte{
+			"api_key": []byte(encodedKey),
+		},
+	}
+}
+
+// IsManagedByAutoOps checks if an API key is managed by the autoops controller.
+func IsManagedByAutoOps(metadata map[string]any) bool {
+	if metadata == nil {
+		return false
+	}
+	_, exists := metadata[PolicyNameLabelKey]
+	return exists
+}
+
+// apiKeyNameFor generates a unique name for the API key according to the policy, and ES instance.
+func apiKeyNameFor(policy autoopsv1alpha1.AutoOpsAgentPolicy, es esv1.Elasticsearch) string {
+	return fmt.Sprintf("autoops-%s-%s-%s-%s", policy.Namespace, policy.Name, es.Namespace, es.Name)
+}
+
+// apiKeySecretNameFrom generates the name for the API key secret from an ES instance.
+func apiKeySecretNameFor(es types.NamespacedName) string {
+	return fmt.Sprintf("%s-%s-%s", es.Name, es.Namespace, autoOpsESAPIKeySecretNameSuffix)
```

Comment:
User namer, to avoid too long secret names.
This is using the wrong namespace. It needs to be in the namespace where the agent is deployed == policy namespace.